rfcs/2022-04-09-accessibility.md

```diff
@@ -0,0 +1,1001 @@
+
+Not communicating dynamic content to users will cause:
+
+- [Failure of Success Criterion 4.1.3](https://www.w3.org/WAI/WCAG21/Techniques/failures/F103) due to providing status messages that cannot be programmatically determined through role or properties
+- Related to: [using ARIA role=alert or Live Regions to Identify Errors](https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html)
+
+#### Best practices for dynamic contents:  <!-- omit in toc -->
+
+- Dynamic content which updates without a page reload should either be a live region or a widget.
+- Use `aria-live` or [roles with implicit live region attributes](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions#roles_with_implicit_live_region_attributes) for dynamic content.
+- Make sure to use `POLITENESS_SETTING` for `aria-live` depending on the criticality of the content.
+
+**aria-live**
+
+`aria-live` attribute can be used to make a live region container. It accepts three values:
+
+| Value           | Description                                                                                                                                                                    |
+| :-------------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `assertive`     | Indicates that updates to the region have the highest priority and should be presented the user immediately.                                                                   |
+| `off (default)` | Indicates that updates to the region should not be presented to the user unless the user is currently focused on that region.                                                  |
+| `polite`        | Indicates that updates to the region should be presented at the next graceful opportunity, such as at the end of speaking the current sentence or when the user pauses typing. |
+
+#### Platform Specific Implementation <!-- omit in toc -->
+
+**Web:**
+
+In web we can simply add `aria-live` to any element or use [roles which have implicit live region attributes](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions#roles_with_implicit_live_region_attributes)
+
+> [Demo](https://codesandbox.io/s/blade-a11y-rfc-arialive-demo-h53z61?file=/src/App.tsx)
+
+https://user-images.githubusercontent.com/35374649/170882365-47b9caa8-eff8-4572-a319-d0f63785f497.mov
+
+We can also use the [@react-aria/live-announcer](https://github.com/adobe/react-spectrum/tree/main/packages/%40react-aria/live-announcer) package which gives us programmatic way of triggering announcements with `announce()` function, by creating temporary live-regions.
```

Comment:
how do we decide whether to use `aria-live` attribute or this utility?

Reply:
So the difference this announce() utility will be handy because this utility will create hidden DOM nodes with VisuallyHidden programatically thus it won't be visible to users. But with aria-live we have to manually make sure it's hidden. 

It's based on our usecases.

Reply:
makes sense. but can't button loading have 

`<button aria-live={isLoading? 'assertive':undefined}>Submit Form</button>

Reply:
It will work, but it's not the correct way to handle live announcements, live announcements should always have their own independent container... putting aria-live directly inside the button you are coupling the announcement to the button's text tightly and there is no way to have different announcement and different text for button. 

You'll need to do this
```ts
<button aria-live={isLoading? 'assertive':undefined}>{isLoading ? "Loading" : "Submit button"}</button>
```

With this example how would you let users know that button has stopped loading? You cannot announce "Finished loading" since if you do the button's text will also say "Finished loading"